funcs/func_groupcount.c

```diff
@@ -273,21 +1266,287 @@ static int group_list_function_read(struct ast_channel *chan, const char *cmd,
+		return -1;
+	}
+
+	ast_verbose("%s\n", line);
+	ast_verbose(FORMAT_STRING_CHANNELS, "Channel", "Group", "Category\n");
+
+	ast_app_group_list_rdlock();
+	gi = ast_app_group_list_head();
+	while (gi) {
+		ast_str_append(&out, 0, FORMAT_STRING_CHANNELS, ast_channel_name(gi->chan), gi->group, (ast_strlen_zero(gi->category) ? "(default)" : gi->category));
+		numchans++;
+		gi = AST_LIST_NEXT(gi, group_list);
+	}
+
+	ast_app_group_list_unlock();
+
+	ast_str_append(&out, 0, "%d active group assignment%s\n", numchans, ESS(numchans));
+
+	ast_str_append(&out, 0, "%s\n", thinline);
+	/****************** Group Variables ******************/
+
+	ast_str_append(&out, 0, "Group	  Variables    Category\n");
+
+	/* Print group variables */
+	ast_app_group_meta_rdlock();
+	gmi = ast_app_group_meta_head();
+	while (gmi) {
+		ast_str_append(&out, 0, FORMAT_STRING_GROUPS, gmi->group, (strcmp(gmi->category, "") ? gmi->category : "(Default)"));
+		numgroups++;
+		headp = &gmi->varshead;
+
+		AST_LIST_TRAVERSE(headp, variable, entries) {
+			ast_str_append(&out, 0, FORMAT_STRING_VAR, ast_var_name(variable), ast_var_value(variable));
+		}
+
+		gmi = AST_LIST_NEXT(gmi, group_meta_list);
+	}
+	ast_app_group_meta_unlock();
+
+	ast_str_append(&out, 0, "%s\n", line);
+
+	ast_verbose("%s\n", ast_str_buffer(out));
+
+	ast_free(out);
+
+	return 0;
+
+#undef FORMAT_STRING_CHANNELS
+#undef FORMAT_STRING_GROUPS
+#undef FORMAT_STRING_VAR
+}
+
+static struct ast_custom_function group_match_list_function = {
+	.name = "GROUP_MATCH_LIST",
+	.read = group_match_list_function_read,
+	.write = NULL,
+};
+
 static struct ast_custom_function group_list_function = {
 	.name = "GROUP_LIST",
 	.read = group_list_function_read,
 	.write = NULL,
 };
 
+static int group_channel_list_function_read(struct ast_channel *chan, const char *cmd, char *data, char *buf, size_t len)
+{
+	char group[MAX_GROUP_LEN] = "";
+	char category[MAX_CATEGORY_LEN] = "";
+	struct ast_group_info *gi = NULL;
+	struct ast_str *out = ast_str_create(1024);
+	int out_len = 0;
+
+	AST_DECLARE_APP_ARGS(args,
+		AST_APP_ARG(groupcategory);
+	);
+	AST_STANDARD_APP_ARGS(args, data);
+
+	buf[0] = 0;
+
+	if (ast_strlen_zero(args.groupcategory)) {
+		ast_log(LOG_WARNING, "Syntax GROUP_CHANNEL_LIST(group[@category])\n");
+		return -1;
+	}
+
+	if (ast_app_group_split_group(args.groupcategory, group, sizeof(group), category, sizeof(category))) {
+		return -1;
+	}
+
+	ast_app_group_list_rdlock();
+	gi = ast_app_group_list_head();
+	while (gi) {
+		if (!strcasecmp(gi->group, group) && (ast_strlen_zero(category) || (!ast_strlen_zero(gi->category) && !strcasecmp(gi->category, category)))) {
+			ast_str_append(&out, 0, "%s,", ast_channel_name(gi->chan));
+		}
+
+		gi = AST_LIST_NEXT(gi, group_list);
+	}
+	ast_app_group_list_unlock();
+
+	out_len = strlen(ast_str_buffer(out));
+
+	if (out_len) {
+		if (out_len <= len) {
+		  len -= 1; /* rid , only if we didn't go over the buffer limit */
+		}
+
+		strncpy(buf, ast_str_buffer(out), len);
```

Comment:
Don't use strncpy, use ast_copy_string.
But don't use ast_copy_string, we already have a buffer, use that directly

Reply:
I think it's just nitpicking here on this comment, for no functional gain.  It sounds like more 'c purist' than actually making any actual improvements in the code itself.

Yes... we do have a buffer to work with, but as usual 'the work has to be done somewhere'.  So either keep track of the write position of *buf with pointer math, or use something that *already* has smarts of position.

Current version:
- build a string, copy it into buf (ever-so-slightly wasteful of a buffer)

What you are suggesting:
- Write a more delicate method to accomplish the same thing... using a buffer directly with... what exactly?  sprintf() or strncat.  Now we need to keep track of where we are and not overwrite the end

new code would be something like...
loop
  strncat, with max of (length - written-so-far).. 
  bump the write pointer for the next strncat
  pos+= additional length written

  and then repeated calls to strncat will of course need the length of what's being copied... 

My vote is go with safe-and-effective string utilities to do what they are designed to do, rather than hand-code delicate string/pointer code to build a string for something that I don't personally really think needs that kind of level of optimization.  And considering we need to keep track of lengths written _anyway_ , there's literally no computational advantage of doing it 'by hand' versus using strings.h because internally, ast_str keeps track of length for you.

Reply:
This one is not nitpicking, there is a big performance difference.
Don't use strncat, that's also a wasteful function. Just use snprintf each time, it's not that complicated.

```
len = snprintf(pos, left, "%s", foobar);
pos += len;
left -= len;
```

snprintf already returns bytes copied for free (unless truncation occurs).

The advantage is that `ast_str` involves allocation, so if you are going to do that, either use the ast_str version as Sean said or do a zero-copy version into the buffer already provided.

Reply:
Ah okay.  Makes sense. No problem.